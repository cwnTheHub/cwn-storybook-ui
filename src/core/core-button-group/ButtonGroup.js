import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Box from "../core-box/Box";
import Text from "../core-text/Text";
import { componentWithName } from "../../util-prop-types";
import { safeRest } from "../../util-helpers";
import ButtonGroupItem from "./ButtonGroupItem/ButtonGroupItem";

const StyledButtonGroup = styled(Box)({
  flexFlow: "row wrap",
  maxWidth: "784px",
});

const ButtonGroup = React.forwardRef(
  (
    {
      name,
      onChange,
      onFocus,
      onBlur,
      value,
      label,
      children,
      readOnly,
      showFieldset,
      ...rest
    },
    ref
  ) => {
    const passedButtons = React.Children.map(children, (child) =>
      React.cloneElement(child, {
        name,
        onChange,
        onFocus,
        onBlur,
        checked:
          typeof value !== "undefined"
            ? value === child.props.value
            : undefined,
        readOnly,
      })
    );

    const buttonValues = [];
    Object.keys(passedButtons).forEach((key) => {
      buttonValues.push(passedButtons[key].props.value);
    });
    if (!showFieldset) {
      return (
        <Box between={2}>
          <Text bold size="medium">
            {label}
          </Text>
          <StyledButtonGroup between={3} inline>
            {passedButtons}
          </StyledButtonGroup>
        </Box>
      );
    }
    return (
      <fieldset {...safeRest(rest)} name={name} ref={ref}>
        <legend>
          <Text bold size="medium">
            {label}
          </Text>
        </legend>

        <StyledButtonGroup between={3} inline>
          {passedButtons}
        </StyledButtonGroup>
      </fieldset>
    );
  }
);
ButtonGroup.displayName = "ButtonGroup";

ButtonGroup.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.string,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  readOnly: PropTypes.bool,
  showFieldset: PropTypes.bool,
  children: componentWithName("ButtonGroup.Item", true).isRequired,
};

ButtonGroup.defaultProps = {
  onFocus: undefined,
  onBlur: undefined,
  onChange: undefined,
  value: undefined,
  readOnly: undefined,
  showFieldset: false,
};

ButtonGroup.Item = ButtonGroupItem;

export default ButtonGroup;
