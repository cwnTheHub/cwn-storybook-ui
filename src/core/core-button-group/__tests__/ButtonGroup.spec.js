import React from "react";
import { shallow, mount } from "enzyme";


import ButtonGroup from "../ButtonGroup";
import { toBeChecked } from "@testing-library/jest-dom/dist/matchers";
import A11yContent from "../../core-a11y-content/A11yContent";

describe("ButtonGroup", () => {
  const defaultProps = {
    name: "storageSize",
    onChange: () => {
      return true;
    },
    value: "64gb",
    label: "Please select a storage size",
  };
  const doShallow = (props = {}) =>
    shallow(
      <ButtonGroup {...defaultProps} {...props}>
        <ButtonGroup.Item value="64gb">64 GB</ButtonGroup.Item>
        <ButtonGroup.Item value="128gb">128 GB</ButtonGroup.Item>
        <ButtonGroup.Item value="256gb">256 GB</ButtonGroup.Item>
      </ButtonGroup>
    );

  const doMount = (props = {}) =>
    mount(
      <ButtonGroup {...defaultProps} {...props}>
        <ButtonGroup.Item value="64gb">64 GB</ButtonGroup.Item>
        <ButtonGroup.Item value="128gb">128 GB</ButtonGroup.Item>
        <ButtonGroup.Item value="256gb">256 GB</ButtonGroup.Item>
      </ButtonGroup>
    );

  it("renders", () => {
    const buttonGroup = doMount();

    expect(buttonGroup).toMatchSnapshot();
  });

  it("selects the correct button", () => {
    const buttonGroup = doMount({ value: "128gb" });
    expect(
      buttonGroup.children().find('input[value="128gb"]').get(0).props.checked
    ).toEqual(true);
  });

  it("renders a group label", () => {
    const buttonGroup = doShallow();
    expect(buttonGroup.find("legend")).toHaveLength(1);
  });

  it("passes additional attributes to the element", () => {
    const buttonGroup = doShallow({
      id: "the-id",
      "data-some-attr": "some value",
    });

    expect(buttonGroup.props().id).toBe("the-id");
    expect(buttonGroup.props()["data-some-attr"]).toBe("some value");
  });

  describe("Accepts A11yContent", () => {
    it("connects to ButtonGroup.Item", () => {
      const buttonGroup = shallow(
        <ButtonGroup {...defaultProps}>
          <ButtonGroup.Item value="64gb">64 GB</ButtonGroup.Item>
          <ButtonGroup.Item value="512gb">
            512 GB
            <A11yContent>test</A11yContent>
          </ButtonGroup.Item>
        </ButtonGroup>
      );

      expect(buttonGroup.contains(<A11yContent>test</A11yContent>)).toBe(true);
    });
  });

  it("does not allow custom CSS", () => {
    const buttonGroup = doShallow({
      className: "my-custom-class",
      style: { color: "hotpink" },
    });

    expect(buttonGroup.props()["className"]).not.toEqual("my-custom-class");
    expect(buttonGroup.props()["style"]).toBe(undefined);
  });
});
