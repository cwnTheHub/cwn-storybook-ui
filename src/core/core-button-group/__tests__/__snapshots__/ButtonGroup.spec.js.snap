// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup renders 1`] = `
<ButtonGroup
  label="Please select a storage size"
  name="storageSize"
  onChange={[Function]}
  showFieldset={true}
  value="64gb"
>
  <fieldset
    name="storageSize"
  >
    <legend>
      <Text
        block={false}
        bold={true}
        invert={false}
        size="medium"
      >
        <styled.span
          block={false}
          bold={true}
          invert={false}
          size="medium"
        >
          <span
            className="sc-gEvEer fKNAcX"
            size="medium"
          >
            Please select a storage size
          </span>
        </styled.span>
      </Text>
    </legend>
    <Styled(Box)
      between={3}
      inline={true}
    >
      <Box
        between={3}
        className="sc-iGgWBj gra-DKZ"
        inline={true}
        tag="div"
      >
        <styled.div
          between={3}
          className="sc-iGgWBj gra-DKZ"
          inline={true}
          tag="div"
        >
          <div
            className="sc-aXZVg dPmMeI sc-iGgWBj gra-DKZ sc-iGgWBj gra-DKZ"
          >
            <ButtonGroup.Item
              checked={true}
              key=".0"
              name="storageSize"
              onChange={[Function]}
              value="64gb"
            >
              <styled.div>
                <div
                  className="sc-eqUAAy bVaFdC"
                >
                  <styled.input
                    checked={true}
                    id="storagesize_64gb"
                    name="storageSize"
                    onChange={[Function]}
                    type="radio"
                    value="64gb"
                  >
                    <input
                      checked={true}
                      className="sc-fqkvVR cciTRq"
                      id="storagesize_64gb"
                      name="storageSize"
                      onChange={[Function]}
                      type="radio"
                      value="64gb"
                    />
                  </styled.input>
                  <styled.label
                    htmlFor="storagesize_64gb"
                  >
                    <label
                      className="sc-dcJsrY lhdwjz"
                      htmlFor="storagesize_64gb"
                    >
                      64 GB
                    </label>
                  </styled.label>
                </div>
              </styled.div>
            </ButtonGroup.Item>
            <ButtonGroup.Item
              checked={false}
              key=".1"
              name="storageSize"
              onChange={[Function]}
              value="128gb"
            >
              <styled.div>
                <div
                  className="sc-eqUAAy bVaFdC"
                >
                  <styled.input
                    checked={false}
                    id="storagesize_128gb"
                    name="storageSize"
                    onChange={[Function]}
                    type="radio"
                    value="128gb"
                  >
                    <input
                      checked={false}
                      className="sc-fqkvVR cciTRq"
                      id="storagesize_128gb"
                      name="storageSize"
                      onChange={[Function]}
                      type="radio"
                      value="128gb"
                    />
                  </styled.input>
                  <styled.label
                    htmlFor="storagesize_128gb"
                  >
                    <label
                      className="sc-dcJsrY lhdwjz"
                      htmlFor="storagesize_128gb"
                    >
                      128 GB
                    </label>
                  </styled.label>
                </div>
              </styled.div>
            </ButtonGroup.Item>
            <ButtonGroup.Item
              checked={false}
              key=".2"
              name="storageSize"
              onChange={[Function]}
              value="256gb"
            >
              <styled.div>
                <div
                  className="sc-eqUAAy bVaFdC"
                >
                  <styled.input
                    checked={false}
                    id="storagesize_256gb"
                    name="storageSize"
                    onChange={[Function]}
                    type="radio"
                    value="256gb"
                  >
                    <input
                      checked={false}
                      className="sc-fqkvVR cciTRq"
                      id="storagesize_256gb"
                      name="storageSize"
                      onChange={[Function]}
                      type="radio"
                      value="256gb"
                    />
                  </styled.input>
                  <styled.label
                    htmlFor="storagesize_256gb"
                  >
                    <label
                      className="sc-dcJsrY lhdwjz"
                      htmlFor="storagesize_256gb"
                    >
                      256 GB
                    </label>
                  </styled.label>
                </div>
              </styled.div>
            </ButtonGroup.Item>
          </div>
        </styled.div>
      </Box>
    </Styled(Box)>
  </fieldset>
</ButtonGroup>
`;
